File: CryptoApp/Shared Components/SkeletonView/SkeletonView.m
//
//  SkeletonView.m
//  CryptoApp
//
//  Created by Jansen Castillo
//

#import "SkeletonView.h"

@interface SkeletonView ()

@property (nonatomic, strong) CAGradientLayer *gradientLayer;
@property (nonatomic, assign) BOOL isAnimating;

@end

@implementation SkeletonView

// MARK: - Initialization

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self setupSkeleton];
    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)coder {
    self = [super initWithCoder:coder];
    if (self) {
        [self setupSkeleton];
    }
    return self;
}

// MARK: - Deallocation

- (void)dealloc {
    [self stopShimmering];
}

// MARK: - Setup

- (void)setupSkeleton {
    self.backgroundColor = [UIColor systemGray5Color];
    self.layer.cornerRadius = 4.0;
    
    // Setup gradient layer for shimmer effect
    self.gradientLayer = [[CAGradientLayer alloc] init];
    self.gradientLayer.colors = @[
        (id)[UIColor systemGray5Color].CGColor,
        (id)[UIColor systemGray4Color].CGColor,
        (id)[UIColor systemGray5Color].CGColor
    ];
    self.gradientLayer.locations = @[@0, @0.5, @1];
    self.gradientLayer.startPoint = CGPointMake(0, 0.5);
    self.gradientLayer.endPoint = CGPointMake(1, 0.5);
    [self.layer addSublayer:self.gradientLayer];
}

- (void)layoutSubviews {
    [super layoutSubviews];
    self.gradientLayer.frame = self.bounds;
}

// MARK: - Animation

- (void)startShimmering {
    if (self.isAnimating) {
        return;
    }
    self.isAnimating = YES;
    
    CABasicAnimation *animation = [CABasicAnimation animationWithKeyPath:@"locations"];
    animation.fromValue = @[@(-1.0), @(-0.5), @(0.0)];
    animation.toValue = @[@(1.0), @(1.5), @(2.0)];
    animation.duration = 1.5;
    animation.repeatCount = INFINITY;
    [self.gradientLayer addAnimation:animation forKey:@"shimmer"];
}

- (void)stopShimmering {
    self.isAnimating = NO;
    if (self.gradientLayer) {
        [self.gradientLayer removeAnimationForKey:@"shimmer"];
    }
}

// MARK: - Convenience Factory Methods

+ (instancetype)textSkeletonWithWidth:(CGFloat)width height:(CGFloat)height {
    SkeletonView *skeleton = [[SkeletonView alloc] init];
    skeleton.translatesAutoresizingMaskIntoConstraints = NO;
    [NSLayoutConstraint activateConstraints:@[
        [skeleton.widthAnchor constraintEqualToConstant:width],
        [skeleton.heightAnchor constraintEqualToConstant:height]
    ]];
    return skeleton;
}

+ (instancetype)circleSkeletonWithDiameter:(CGFloat)diameter {
    SkeletonView *skeleton = [[SkeletonView alloc] init];
    skeleton.translatesAutoresizingMaskIntoConstraints = NO;
    skeleton.layer.cornerRadius = diameter / 2.0;
    [NSLayoutConstraint activateConstraints:@[
        [skeleton.widthAnchor constraintEqualToConstant:diameter],
        [skeleton.heightAnchor constraintEqualToConstant:diameter]
    ]];
    return skeleton;
}

+ (instancetype)rectangleSkeletonWithWidth:(CGFloat)width height:(CGFloat)height cornerRadius:(CGFloat)cornerRadius {
    SkeletonView *skeleton = [[SkeletonView alloc] init];
    skeleton.translatesAutoresizingMaskIntoConstraints = NO;
    skeleton.layer.cornerRadius = cornerRadius;
    [NSLayoutConstraint activateConstraints:@[
        [skeleton.widthAnchor constraintEqualToConstant:width],
        [skeleton.heightAnchor constraintEqualToConstant:height]
    ]];
    return skeleton;
}

+ (instancetype)resizableSkeletonWithCornerRadius:(CGFloat)cornerRadius {
    SkeletonView *skeleton = [[SkeletonView alloc] init];
    skeleton.translatesAutoresizingMaskIntoConstraints = NO;
    skeleton.layer.cornerRadius = cornerRadius;
    return skeleton;
}

@end

// MARK: - Container View for Multiple Skeletons

@interface SkeletonContainerView ()

@property (nonatomic, strong) NSMutableArray<SkeletonView *> *skeletonViews;

@end

@implementation SkeletonContainerView

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        self.skeletonViews = [[NSMutableArray alloc] init];
    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)coder {
    self = [super initWithCoder:coder];
    if (self) {
        self.skeletonViews = [[NSMutableArray alloc] init];
    }
    return self;
}

// MARK: - Deallocation

- (void)dealloc {
    [self stopShimmering];
    [self removeAllSkeletons];
}

- (void)addSkeletonViews:(NSArray<SkeletonView *> *)views {
    [self.skeletonViews addObjectsFromArray:views];
    for (SkeletonView *view in views) {
        [self addSubview:view];
    }
}

- (void)startShimmering {
    for (SkeletonView *skeletonView in self.skeletonViews) {
        [skeletonView startShimmering];
    }
}

- (void)stopShimmering {
    for (SkeletonView *skeletonView in self.skeletonViews) {
        [skeletonView stopShimmering];
    }
}

- (void)removeAllSkeletons {
    if (self.skeletonViews) {
        for (SkeletonView *skeletonView in self.skeletonViews) {
            [skeletonView stopShimmering];
            [skeletonView removeFromSuperview];
        }
        [self.skeletonViews removeAllObjects];
    }
}

@end

File: CryptoApp/Shared Components/SkeletonView/SkeletonView.swift
import UIKit

final class SkeletonView: UIView {
    
    // MARK: - Properties
    
    private let gradientLayer = CAGradientLayer()
    private var isAnimating = false
    
    // MARK: - Initialization
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setupSkeleton()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        setupSkeleton()
    }
    
    // MARK: - Setup
    
    private func setupSkeleton() {
        backgroundColor = UIColor.systemGray5
        layer.cornerRadius = 4
        
        // Setup gradient layer for shimmer effect
        gradientLayer.colors = [
            UIColor.systemGray5.cgColor,
            UIColor.systemGray4.cgColor,
            UIColor.systemGray5.cgColor
        ]
        gradientLayer.locations = [0, 0.5, 1]
        gradientLayer.startPoint = CGPoint(x: 0, y: 0.5)
        gradientLayer.endPoint = CGPoint(x: 1, y: 0.5)
        layer.addSublayer(gradientLayer)
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        gradientLayer.frame = bounds
    }
    
    // MARK: - Animation
    
    func startShimmering() {
        guard !isAnimating else { return }
        isAnimating = true
        
        let animation = CABasicAnimation(keyPath: "locations")
        animation.fromValue = [-1.0, -0.5, 0.0]
        animation.toValue = [1.0, 1.5, 2.0]
        animation.duration = 1.5
        animation.repeatCount = .infinity
        gradientLayer.add(animation, forKey: "shimmer")
    }
    
    func stopShimmering() {
        isAnimating = false
        gradientLayer.removeAnimation(forKey: "shimmer")
    }
}

// MARK: - Convenience Factory Methods

extension SkeletonView {
    
    /// Creates a skeleton view that mimics a text label
    static func textSkeleton(width: CGFloat, height: CGFloat = 16) -> SkeletonView {
        let skeleton = SkeletonView()
        skeleton.translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            skeleton.widthAnchor.constraint(equalToConstant: width),
            skeleton.heightAnchor.constraint(equalToConstant: height)
        ])
        return skeleton
    }
    
    /// Creates a skeleton view that mimics a circular image
    static func circleSkeleton(diameter: CGFloat) -> SkeletonView {
        let skeleton = SkeletonView()
        skeleton.translatesAutoresizingMaskIntoConstraints = false
        skeleton.layer.cornerRadius = diameter / 2
        NSLayoutConstraint.activate([
            skeleton.widthAnchor.constraint(equalToConstant: diameter),
            skeleton.heightAnchor.constraint(equalToConstant: diameter)
        ])
        return skeleton
    }
    
    /// Creates a skeleton view that mimics a rectangular area
    static func rectangleSkeleton(width: CGFloat, height: CGFloat, cornerRadius: CGFloat = 4) -> SkeletonView {
        let skeleton = SkeletonView()
        skeleton.translatesAutoresizingMaskIntoConstraints = false
        skeleton.layer.cornerRadius = cornerRadius
        NSLayoutConstraint.activate([
            skeleton.widthAnchor.constraint(equalToConstant: width),
            skeleton.heightAnchor.constraint(equalToConstant: height)
        ])
        return skeleton
    }
    
    /// Creates a skeleton view that will be sized by its container constraints
    static func resizableSkeleton(cornerRadius: CGFloat = 4) -> SkeletonView {
        let skeleton = SkeletonView()
        skeleton.translatesAutoresizingMaskIntoConstraints = false
        skeleton.layer.cornerRadius = cornerRadius
        return skeleton
    }
}

// MARK: - Container View for Multiple Skeletons

final class SkeletonContainerView: UIView {
    
    private var skeletonViews: [SkeletonView] = []
    
    func addSkeletonViews(_ views: [SkeletonView]) {
        skeletonViews.append(contentsOf: views)
        views.forEach { addSubview($0) }
    }
    
    func startShimmering() {
        skeletonViews.forEach { $0.startShimmering() }
    }
    
    func stopShimmering() {
        skeletonViews.forEach { $0.stopShimmering() }
    }
    
    func removeAllSkeletons() {
        skeletonViews.forEach { $0.removeFromSuperview() }
        skeletonViews.removeAll()
    }
} 